[B+C] Add MetadataProvider for on-demand metadata [BUKKIT-3625]

crast opened the pull request:
At the very core of the metadata system is the idea of asking a question. A plugin asks a question, like “What is this player’s chat prefix?” and some other plugin is able to provide an answer. 

The way it’s currently designed, for some plugin (say a chat plugin) to ask what a prefix is, some other plugin (say a permissions plugin or user manager) would have to set the "chatPrefix" key on every user before it's asked. (say when they log in.) This isn’t very effective, and it isn’t quite as “lazy” as one would hope, what if the value is never asked for? Then we've done all that work for nothing.

What would be useful, instead, if the metadata system provided a way for a plugin to register a metadata ‘provider’ for a key. Calling <object>.getMetadata(key) would query the provider if no value is set, passing it the object, for an appropriate metadata value.

To use the chatPrefix example from before, MyAwesomePermissions would register a provider for chatPrefix. Then at some point, BobsChatPlugin calls player.getMetadata("chatPrefix") which asks the provider for the chatPrefix. This also means if the chatPrefix is never requested (say you don't have a chat plugin to use it), there's no work done or memory used in creating the values.

**More Info**
full description at https://bukkit.atlassian.net/browse/BUKKIT-3625
Also see companion pull request https://github.com/Bukkit/CraftBukkit/pull/1032

**Full list of changes**
    \* Created provider interface
    \* Added provider registration to MetadataStore
    \* Added unit tests for provider usage
    \* Updated bukkit.Server interface with methods for registering
    Player/Entity/World providers
    \* Updated bukkit.Bukkit with static implementations
    \* Updated World with way to register Block providers

crast committed:
Add MetadataProvider for on-demand metadata

    * Created provider interface
    * Added provider registration to MetadataStore
    * Added unit tests for provider usage
    * Updated bukkit.Server interface with methods for registering
    Player/Entity/World providers
    * Updated bukkit.Bukkit with static implementations
    * Updated World with way to register Block providers

YLivay commented:
This sounds really good!

Wolvereness commented:
I really like the idea of meta data provider. However, what I would like to see from this idea:
- Registration is based on MetaData-able .class. We don't need methods everywhere for every type of class. Two methods, one for registration, one for unregistration. Less is more (when intuitive).
- Don't require a specific token word for each registration. Let the plugin generally handle any token another plugin attempts to access. Register once, and let it determine if the incoming token will have data.
- Sanitization of calls to the plugin. We should (almost) expect an exception being fired from a provider.

crast commented:
@Wolvereness 
1. This raises a few problems. There were two ways I thought to interpret what you said:
   1. maybe you meant add a pair of methods to Metadatable interface: The problem there is if Player is metadatable, then you would need an instance to a specific player to register the provider, and the provider is technically for all players, not a single player, and interfaces can't define static-methods, only instance-methods.
   2. or maybe you meant to have a method like e.g. `registerMetadataProvider(MetadataProvider<T> provider, ...)` on e.g. Server and Bukkit. While this sounds good at the start, the main problem is type erasure would force the method to do reflection to figure out what sort of provider this is to begin with, and it gets very hairy very fast.  Or alternately you'd need to do e.g. `registerMetadataProvider(Player.class, MetadataProvider<Player>)` which again requires quite a bit of reflection magics.
      In this case I feel like the individual methods makes things clear and type-safe, even though it does introduce more methods. It also handles the scoping, e.g. block metadata providers being registered per-world, while player metadata is server-wide, which a single registration method would need to be kludgy to deal with.
2. on token words: I guess I'm somewhat torn on this. Because Metadata does not allow iterating a list of the metadata names on an object (you need to know the key) and also the general way it's implemented, metadata makes the most sense for checking known keys, and specifically it's designed for plugin interop.  Having to register a provider on a given key (e.g. chatPrefix) allows one to avoid the likely scenario of iterating through a whole bunch of providers, none of which have any interest in the key in question.
3. What sort of exception would you want to be thrown? I feel like declaring a generic exception throws would be simply asking for poor too-generic exception handling code in the case of the metadata storage, with no real tangible benefit, as it will further confuse users who won't know what exactly to use the exception for either.

Wolvereness commented:
1.1: My specific point is not to add methods to specific metadataable types, but rather to the over-arching system.
1.2: Type erasure/generics makes this easy. `public <T extends Metadataable> void registerMetadataProvider(Provider<T> provider, Class<T> clazz)`. That doesn't require reflection at all. Incoming object is checked against the provided class `if (clazz.isInstance(object)) {...}`. To call into the provider, you just need `clazz.cast(object)` to restore the generic type and maintain runtime safety. This also allows for the provider interface to have generics on all signatures.

2: Events... We already iterate through plugins and this solution wouldn't even be using reflection.
3: Not the point, the point is that we need `try {} catch {}` blocks everywhere we call into another plugin.

riking commented on src/main/java/org/bukkit/metadata/MetadataStoreBase.java:
```diff
@@ -55,12 +56,16 @@ public synchronized void setMetadata(T subject, String metadataKey, MetadataValu
      * @see MetadataStore#getMetadata(Object, String)
      */
     public synchronized List<MetadataValue> getMetadata(T subject, String metadataKey) {
+       
```
whitespace! :(

riking commented on src/main/java/org/bukkit/metadata/MetadataStoreBase.java:
```diff
@@ -152,4 +184,20 @@ private String cachedDisambiguate(T subject, String metadataKey) {
      */
     protected abstract String disambiguate(T subject, String metadataKey);
+    
+    
```
more whitespace

crast commented:
@Wolvereness 
In response to 1.2, how do you then register the metadata provider for Block? (There's a different block metadata store per World) Currently I have that implemented on World so that it can be done in the appropriate scope.

crast committed:
Cleaned up whitespace.

crast committed:
Catch when a metadata provider fails.

Catch the exception and log it to the server's logger.

crast committed:
Remove redundant getServer()

crast commented:
(Message removed after discussing with Wolvereness)

crast committed:
Implement generic registration of metadata providers.

Also take provider registry out of the MetadatadaStore interface, and
delegate this to the server to figure out on the CraftBukkit side,
hiding implementation details.

crast committed:
Remove provider registration code

crast committed:
Make OfflinePlayer metadatable

crast committed:
Complete provider revamp.

This is based on Amaranth/Wolvereness's concepts to make the Provider
system into a dispatch system similar in idea to events, only sans
reflection stuff.
* Provider loses the getOwningPlugin, to make implementation easier.
Instead of getOwningPlugin, registration gets a plugin to the
parameter, similar to registering events, scheduler tasks, etc.
* Provider registration is not done on a contextual key, instead
there's one per scope/plugin.
* Metadata Store base implementation now expects a collection of
providers, not a map of key=>providers and will run multiple providers.

crast committed:
Clean up imports and dead test code.

crast commented:
Updated provider system substantially, this now looks more like a dispatch system, with formal registration for providers per-plugin per-scope, and such. Also see companion pull request Bukkit/CraftBukkit#1032

crast committed:
Make MetadataStore and MetadataProvider <T extends Metadatable>

This allows for better generic guarantees.

crast committed:
Implement per-class canonical provider map

crast commented:
Superceded by #845
